torchsynth/synth.py

```diff
@@ -163,9 +173,111 @@ def _forward(self) -> T:
         pitch_envelope = self.pitch_adsr.forward(note_on_duration)
         amp_envelope = self.amp_adsr.forward(note_on_duration)
 
-        audio_out = self.vco_1.forward(midi_f0, pitch_envelope)
+        vco_1_out = self.vco_1.forward(midi_f0, pitch_envelope)
         vco_2_out = self.vco_2.forward(midi_f0, pitch_envelope)
-        audio_out = util.crossfade2D(audio_out, vco_2_out, self.vco_ratio.p("ratio"))
+        audio_out = util.crossfade2D(vco_1_out, vco_2_out, self.vco_ratio.p("ratio"))
 
         audio_out = self.noise.forward(audio_out)
         return self.vca.forward(amp_envelope, audio_out)
+
+
+class FmOperator(AbstractSynth):
+    """
+    An FM operator - a single unit containing a sine wave oscillator that can receive
+    an audio rate frequency modulation signal and contains an amplitude envelope
+    """
+
+    def __init__(self, synthglobals: SynthGlobals, *args, **kwargs):
+        AbstractSynth.__init__(self, synthglobals=synthglobals, *args, **kwargs)
+
+        # Register all modules as children
+        self.add_synth_modules(
+            [
+                ("osc", FmVCO(synthglobals)),
+                ("env", ADSR(synthglobals)),
+                ("amp", VCA(synthglobals)),
+            ]
+        )
+
+    def _forward(self, midi_f0: T, duration: T, modulation: Signal) -> Signal:
+        output = self.osc(midi_f0, modulation)
+        env = self.env(duration)
+        return self.amp(env, output)
+
+
+class FmVoice(AbstractSynth):
+    """
+    A four operator FM synth.
+
+    TODO: Add an LFO and maybe a pitch envelope? Try to implement DX7 style envs?
+    """
+
+    def __init__(self, synthglobals: SynthGlobals, *args, **kwargs):
+        AbstractSynth.__init__(self, synthglobals=synthglobals, *args, **kwargs)
+
+        # Register children SynthModules
+        self.add_synth_modules(
+            [
+                ("keyboard", MonophonicKeyboard(synthglobals)),
+                ("modulation", FmControl(synthglobals)),
+                ("op2_inputs", NormalizedFaderBank(synthglobals, [0.5])),
+                ("op3_inputs", NormalizedFaderBank(synthglobals, [0.25, 0.5])),
+                ("op4_inputs", NormalizedFaderBank(synthglobals, [0.25, 0.33, 0.5])),
+                ("mixer", SoftModeSelector(synthglobals, n_modes=4)),
+            ]
+        )
+
```

Comment:
So we really haven't thought this thru much. i.e. nesting Synths.

I'm concerned here that:
a) randomize won't hit all the sub-synth parameters
b) listing out all the parameter ranges will be broken

Reply:
We're definitely going to want to have stuff nest IMO. PyTorch is quite smart about nesting modules and making sure that parameters are organized nicely within modules -- I tested pretty thoroughly, but would be happy to do some automated stuff on this as well.